### PR TITLE
Feature/consult profile

### PR DESCRIPTION
**Pull request by Cyriacph**
# ✨ feat(profile): add public profile screen and navigation

- #198 
- #239 

---
## 📌 **Summary**
This PR adds a new public profile screen with navigation support, including UI components, tests, and backend integration.

---
### **New Files Added**
- `PublicProfile.kt`: Data class for public profile information.
- `PublicProfileScreen.kt`: Main screen for displaying public profiles.
- `PublicProfileViewModel.kt`: ViewModel for managing public profile data.
- `PublicProfileTestTags.kt`: Test tags for UI testing.
- `PublicProfileTests.kt`: UI tests for the public profile screen.

### **Navigation Changes**
- Added `PublicProfile` screen to navigation graph.
- Updated `NavigationActions.kt` to support navigation to public profiles.
- Modified `RequestList.kt` to navigate to public profiles from request items.

### **UI Components**
- `FollowButton`: Button for following/unfollowing users.
- `PublicProfileHeader`: Header component for public profiles.
- `PublicProfileScreen`: Main screen component for public profiles.

### **Tests**
- Added UI tests for public profile screen and components.
- Updated navigation tests to include public profile screen.

---
## 🧪 **Testing Details**

| Component          | Test Coverage                     | Verification Method          |
|--------------------|-----------------------------------|------------------------------|
| Public Profile Screen | UI rendering and functionality   | Compose UI tests             |
| Navigation         | Navigation to public profile     | Navigation tests             |
| Follow Button      | Follow/unfollow functionality    | UI tests                     |

---
### **Public Profile Screen**
- **Before**: No public profile screen.
- **After**: Full-featured public profile screen with follow button and user information.
- **Impact**: Enhanced user experience with access to public profiles.

---
## 📸 **Screenshots**
image

image

---
## 🚀 **Impact & Benefits**
✅ **New Public Profile Feature** - Users can view public profiles with detailed information.
✅ **Improved Navigation** - Seamless navigation from request list to public profiles.
✅ **Comprehensive Test Coverage** - UI and navigation tests ensure reliability.

---
## 📝 **Notes & Context**
- **Testing**: Run UI tests and navigation tests to verify functionality.
- **Compatibility**: Ensure all new features integrate smoothly with existing code.
- **Future Work**: Potential enhancements to profile search and image handling.
---
## Warning!
This code implementation and pull request were done with the help of an AI Assistant tool.
---
## **Review Checklist**
- [ ] Verify public profile screen renders correctly.
- [ ] Test navigation to public profile from request list.
- [ ] Check follow/unfollow functionality.
- [ ] Run all tests to ensure coverage and reliability.

### CODE REVIEW

File: app/src/androidTest/java/com/android/sample/navigation/NavigationTests.kt
Comment (Thibaud-Babin) [inline app/src/androidTest/java/com/android/sample/navigation/NavigationTests.kt]: small modification added, the test you added seems good to me

File: app/src/main/java/com/android/sample/ui/navigation/NavigationScreen.kt
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/navigation/NavigationScreen.kt]: This code effectively sets up navigation for a PublicProfile screen in Jetpack Compose, with proper argument handling for userId, scoped ViewModel creation, and side-effect management using LaunchedEffect. It’s clean, concise, and follows best practices. However, you should maybe extract the magic values to constnts, and maybe  consider adding error handling for userId or loadPublicProfile failures for completeness.

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfile.kt
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfile.kt]: This data class is simple enough and the vals have clear understandable name. If I add to recommend something would be to add documentation to follow best practice.

File: app/src/main/java/com/android/sample/ui/request/RequestList.kt [hunk 1]
```diff
@@ -120,6 +121,7 @@ fun RequestListScreen(
   val displayed by searchFilterViewModel.displayedRequests.collectAsState()
   val searchQuery by searchFilterViewModel.searchQuery.collectAsState()
   val isSearching by searchFilterViewModel.isSearching.collectAsState()
+  val currentUserId = FirebaseAuth.getInstance().currentUser?.uid
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/request/RequestList.kt]: Nice val declaration, this is the perfect way to do it (+ 1 kudos if you see what I mean)

File: app/src/main/java/com/android/sample/ui/request/RequestList.kt
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/request/RequestList.kt]: The addition of onProfileClick callbacks and consistent use of Modifier.fillMaxSize() enhance functionality and maintainability. While the modifications are practical and improve interactivity, I would only give as an advice to ensure null safety for navigationActions and test edge cases where currentUserId might be null. The updates are concise and follow good practices, but keep an eye on potential code duplication in parameter declaration

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    // Implementation to load public profile from repository
+    // and update _uiState accordingly
+    viewModelScope.launch {
+      _uiState.value = PublicProfileUiState(isLoading = true, errorMessage = null)
+      try {
+        val up = withContext(Dispatchers.IO) { repository.getUserProfile(profileId) }
+        val public = userProfileToPublic(up)
+        _uiState.value =
+            _uiState.value.copy(isLoading = false, profile = public, errorMessage = null)
+      } catch (e: Exception) {
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileViewModel.kt]: > [!TIP]
> The error message in the catch block defaults to a generic message if e.message is null. Consider providing a more           user-friendly or specific error message based on the type of exception.

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        _uiState.value =
+            _uiState.value.copy(
+                isLoading = false,
+                profile = null,
+                errorMessage = e.message ?: "Failed to load profile")
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileViewModel.kt]: > [!CAUTION]
> MAGIC VALUE WARNING CALL THE FREAKIN ARMY. But more seriously yeah consider moving magic values same speech as usual

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileTestTags.kt
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileTestTags.kt]: The separation of the test tags to a different file and the use of them in itself is really good practice, nice job !

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      ProfileState(
+          userName = "Unknown",
+          userEmail = "",
+          profileId = "",
+          kudosReceived = 0,
+          helpReceived = 0,
+          followers = 0,
+          following = 0,
+          arrivalDate = "",
+          userSection = "",
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt]: > [!WARNING]
>  Magic values -> extract it

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+// Placeholder FollowButton composable
+@Composable
+fun FollowButton(isFollowing: Boolean, onToggle: () -> Unit, modifier: Modifier = Modifier) {
+  ElevatedButton(
+      onClick = onToggle,
+      modifier = modifier.size(width = UiDimens.IconMedium * 3, height = UiDimens.IconMedium)) {
+        if (isFollowing) {
+          Text(text = "Unfollow")
+        } else {
+          Text(text = "Follow")
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt]: > [!WARNING]
> Magic values -> extract it

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    }
+  }
+}
+
+// Placeholder FollowButton composable
+@Composable
+fun FollowButton(isFollowing: Boolean, onToggle: () -> Unit, modifier: Modifier = Modifier) {
+  ElevatedButton(
+      onClick = onToggle,
+      modifier = modifier.size(width = UiDimens.IconMedium * 3, height = UiDimens.IconMedium)) {
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt]: > [!WARNING]
> Magic value -> extract it

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+          modifier
+              .fillMaxWidth()
+              .padding(ProfileDimens.HeaderPadding)
+              .testTag(PublicProfileTestTags.PUBLIC_PROFILE_HEADER),
+      colors = CardDefaults.cardColors(containerColor = accent),
+      elevation = CardDefaults.cardElevation(defaultElevation = ProfileDimens.CardElevation)) {
+        Box(modifier = Modifier.padding(ProfileDimens.HeaderPadding)) {
+          Row(verticalAlignment = Alignment.CenterVertically) {
+            ProfilePicture(
+                profileId = state.profile?.userId ?: "",
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt]: > [!WARNING]
> Magic value -> extract it

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+                  style = MaterialTheme.typography.titleMedium,
+                  color = textColor,
+                  modifier = Modifier.testTag(PublicProfileTestTags.PUBLIC_PROFILE_HEADER_NAME))
+              Text(
+                  text = state.profile?.section ?: "None",
+                  style = MaterialTheme.typography.bodyMedium,
+                  color = textColor,
+                  modifier = Modifier.testTag(PublicProfileTestTags.PUBLIC_PROFILE_HEADER_EMAIL))
+            }
+            Spacer(modifier = Modifier.weight(1f))
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt]: > [!WARNING]
> Magic value -> extract it

File: app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/profile/publicProfile/PublicProfileScreen.kt]: The implementation is mostly good and well implemented, you're code is crispy as usual and you used a lot of sexy lambdas. However, this was kind of magic values feast so I would strongly advise moving the constants to a file as there is a lot of repetition of them and maintaing is made tedious by this. Additionally The following mutable might be problematic as mentioned in the comment in the file as if the screen recomposes the state might be set to false  and the UI status won't reflect the status

File: app/src/main/java/com/android/sample/ui/request/RequestList.kt [hunk 9]
```diff
@@ -123,6 +126,23 @@ fun RequestListScreen(
...
+
+  // Centralized profile click handler: handles null currentUserId and owner vs public routing
+  val onProfileClickHandler: OnProfileClick = { id ->
+    if (id.isNotEmpty() && id == currentUserId) {
+      safeNav.navigateTo(Screen.Profile(id))
+    } else {
+      safeNav.navigateTo(Screen.PublicProfile(id))
+    }
+  }
+
```
Comment (Thibaud-Babin) [inline app/src/main/java/com/android/sample/ui/request/RequestList.kt]: Thi i really good !

**Review by Thibaud-Babin — APPROVED** [review]
I see that you added applied hat I recommended by extracting the magic values, adding documentation and taking into account the problematic value that might invalidate the folloing upon recompoe. You also did all of thie very thoroughly !
However, there are still two sonarcloud issues:
image
I will approve this as the most critical problems have been addressed and I'll let you address the sonarcloud issues, Harri might help you with it
